src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AmendReferralService.kt

```diff
@@ -49,7 +64,70 @@ class AmendReferralService(
     referralRepository.save(referral)
     referralEventPublisher.referralComplexityChangedEvent(referral)
 
-    return update
+    return changelog
+  }
+
+  fun updateReferralDesiredOutcomes(
+    referral: Referral,
+    amendDesiredOutcomesDTO: AmendDesiredOutcomesDTO,
+    authentication: JwtAuthenticationToken,
+    serviceCategoryId: UUID
+  ): Changelog? {
+
+    val desiredOutcomeIds = amendDesiredOutcomesDTO.desiredOutcomesIds
+
+    if (desiredOutcomeIds.isEmpty()) {
+      throw ServerWebInputException("desired outcomes cannot be empty")
+    }
+
+    if (referral.approvedActionPlan != null) {
+      throw ServerWebInputException("desired outcomes cannot be updated: the action plan is already approved")
+    }
+
+    if (referral.selectedServiceCategories.isNullOrEmpty()) {
+      throw ServerWebInputException("desired outcomes cannot be updated: no service categories selected for this referral")
+    }
+
+    if (!referral.selectedServiceCategories!!.map { it.id }.contains(serviceCategoryId)) {
+      throw ServerWebInputException("desired outcomes cannot be updated: specified service category not selected for this referral")
+    }
+
+    val validDesiredOutcomeIds = serviceCategoryRepository.findByIdOrNull(serviceCategoryId)?.desiredOutcomes?.map { it.id }
+      ?: throw ServerWebInputException("desired outcomes cannot be updated: specified service category not found")
+
+    if (!validDesiredOutcomeIds.containsAll(desiredOutcomeIds)) {
+      throw ServerWebInputException("desired outcomes cannot be updated: at least one desired outcome is not valid for this service category")
+    }
+
+    if (referral.selectedDesiredOutcomes == null) {
+      referral.selectedDesiredOutcomes = mutableListOf()
+    }
+
+    val actor = userMapper.fromToken(authentication)
+    val oldValue = ReferralAmendmentDetails(values = referral.selectedDesiredOutcomes!!.map { it.desiredOutcomeId.toString() }.toList())
+    val newValue = ReferralAmendmentDetails(values = desiredOutcomeIds.map { it.toString() }.toList())
+
+    referral.selectedDesiredOutcomes!!.removeIf { desiredOutcome -> desiredOutcome.serviceCategoryId == serviceCategoryId }
+    desiredOutcomeIds.forEach { desiredOutcomeId ->
+      referral.selectedDesiredOutcomes!!.add(
+        SelectedDesiredOutcomesMapping(serviceCategoryId, desiredOutcomeId)
+      )
+    }
+
+    val changelog = Changelog(
+      referral.id,
+      UUID.randomUUID(),
+      AmendTopic.DESIRED_OUTCOMES,
+      oldValue,
+      newValue,
+      amendDesiredOutcomesDTO.reasonForChange,
+      OffsetDateTime.now(),
+      actor
+    )
+    changelogRepository.save(changelog)
+    val savedReferral = referralRepository.save(referral)
+    referralEventPublisher.referralDesiredOutcomesChangedEvent(savedReferral)
+    return changelog
```

Comment:
should we be returning the changelog entry?  The user of the endpoint is trying to update the referral, the changelog entry is created as a side effect of this update, so that isn't what I'd expect to see sent back

Reply:
we are not returning the changelog back to the user. It just returns back to the controller. We then return this object

Reply:
I don't need any of the values for the front end, it doesn't use the return values anyway.  
But for a user of this endpoint I don't think that would be an expected response - they are trying to amend a referral, the changelog details are created as a side effect so I wouldn't expect that to be the response (and if they are hitting amend-desired-outcomes why do they need amendTopic in the response?)

Reply:
Though its extracting from the changelog, its actually the value that is been updated that has been passed back. I agree with you saying we don't need amendTopic but if that is case, we don't need other values also to be send back. Instead we can simply send a string message "Details Amended" with a 204 status code or else send back the ReferralDTO object.